WIP - Setup wet run testing

Dhananjhay opened the pull request:
Summary
This PR introduces a fully-automated, label-driven GitHub Actions workflow that:
- Spins up a new ephemeral OpenStack VM on Compute Canada (Arbutus)
- Registers it as a self-hosted GitHub Actions runner
- Waits for the runner to come online
- Downloads wet run test data
- Executes the Hippunfold “wet-run” test inside a Miniforge + Mamba environment
- Tears down the VM and deregisters the runner (success or failure)
- Deletes the runner registration record from GitHub

The workflow is triggered only when the PR is labeled `ready-for-wet-run`. 

To check/do:
- Concurrency amongst different PRs
- Update wet run tests:
1. Side-by-side of the QC results
2. Distribution of vertex-vertex euclidean distances
- Caching Conda environments or pre-baking them into a VM image can significantly speed up repeated runs. However, there are two key bottlenecks:

    Persistent volume mounting:
    The ideal setup would involve storing cached Conda environments on a persistent volume and mounting it during each run. But due to strict volume limits (e.g., a maximum of two attached volumes per instance), this approach does not scale well when running multiple ephemeral runners in parallel.

    Pre-baked Conda environments in the base VM image:
    While this avoids external volume limitations, it bloats the base image, resulting in slower VM boot times and increased snapshot size. This adds overhead during runner startup and can even cause timeouts in CI workflows.
- Use GPUs to accelerate the pipeline

Dhananjhay committed:
add logic for wet run testing on cloud VM ephemeral runners

Dhananjhay committed:
use secret instead of password to login

Dhananjhay committed:
correct secrets spelling

Dhananjhay committed:
use pull_request_target

Dhananjhay committed:
undo last commit and just add pull_request_target logic

Dhananjhay committed:
rectify repo name logic

Dhananjhay committed:
rectify repo name logic

Dhananjhay committed:
change api endpoint to get registration token

Dhananjhay committed:
debug registration token step

Dhananjhay committed:
remove pull_request trigger

Dhananjhay committed:
remove deprecated set-output logic

Dhananjhay committed:
print event trigger name and add permissions

Dhananjhay committed:
replace pull_request_target with pull_request

Dhananjhay committed:
add branches in pull_request_target

Dhananjhay committed:
change to pull_request_target

Dhananjhay committed:
remove branches from pull_request_target

Dhananjhay committed:
replace GITHUB token with PAT

Dhananjhay committed:
stop using pull_request_target

Dhananjhay committed:
validate PAT

Dhananjhay committed:
use new PAT

Dhananjhay committed:
swap token with pat

Dhananjhay committed:
fix cloud-init.yml and inspect cloud-init.generated.yml

Dhananjhay committed:
rectify env subs

Dhananjhay committed:
use 60gb mem instead of 120gb

Dhananjhay committed:
log cloud-init

akhanf committed:
Fixes bug where specific rules were expanding over multiple subjects erroneously (#497)

* fix for expand in two rules, was expanding over multiple subjects
* add test case for multiple subjects

akhanf committed:
fix atlas plugin to be semi-functional when there is no internet connectivity (#494)

* give warning or error depending on internet connectivity

if atlas repo exists, will just warn that updating will be skipped. if
it doesn't exist, will error out..

* fix for pulling updated atlases

Dhananjhay committed:
transfer sudo logic from start.sh to runcmd

Dhananjhay committed:
export runner variable

Dhananjhay committed:
fix github repo name

Dhananjhay committed:
fix runner's name

Dhananjhay committed:
configure miniforge, run dry run, and correctly delete vm after

Dhananjhay committed:
load bashrc and install openstack client

Dhananjhay committed:
add download data logic and remove registration of runner from github ui

Dhananjhay committed:
redirect hippunfold cache to /mnt and fix pr number variable

Dhananjhay committed:
redirect hippunfold cache to /mnt

Dhananjhay committed:
remove manual dispatch

Dhananjhay commented:
Hey all @akhanf @jordandekraker @mackenziesnyder! I was able to orchestrate an entire end to end wet run testing pipeline and I'm excited to share a working example with you all! There's still work to be done but let me know if anyone has any suggestions or feedback!

jordandekraker commented:
Looks awesome, and good to start with 1 lowRes sample for testing. I suppose the next step would be to view a side-by-side of the QC results with the `main` or target of a given PR. Do you know a good way to do this? In addition to that, we could also check the distribution of vertex-vertex euclidean distances (this could be a new PR though)

Dhananjhay commented:
Sounds good. I don't know the answer to that from the top of my head, but I'll add that to the todo list agenda and research on it. Thank you!

Dhananjhay committed:
use a pre-baked image with conda and test data

Dhananjhay committed:
rectify path to snakebids-envs

Dhananjhay committed:
remove trigger on ready-for-wet-run label

Dhananjhay committed:
combine activation + execution in one step

Dhananjhay committed:
check if test data is accessible

Dhananjhay committed:
change data path from /mnt to /opt

Dhananjhay committed:
cache conda envs and wet run data

Dhananjhay committed:
change name of the base iamge

Dhananjhay committed:
change /conda-envs to /data

Dhananjhay committed:
rectify mount source

Dhananjhay committed:
save conda cache and data to snapshot volume

Dhananjhay committed:
use default volume type

Dhananjhay committed:
mount volume after runner registration

Dhananjhay committed:
remove mounting volume logic and bake in the caches conda envs in /

Dhananjhay committed:
increase time for runner registration

Dhananjhay committed:
remove base image logic

Dhananjhay committed:
remove miniforge activation logic

Dhananjhay committed:
redirect the conda-envs to /mnt volume

akhanf committed:
fixes the subfield-tissue jitter in volumetric dseg (#500)

imposes the original gm mask, using the surf-vol labels + dentate with a
nearest voxel approach to fill in the gaps

akhanf committed:
fix CLI arg parsing for laminar_coords_res (#506)

the destination variable wasn't being populated so the argument had no
effect

jordandekraker committed:
updated approach for open isosurface creation (#503)

* no coords masking; instead remove outersurf by removing the largest connected component of outside (invalid) vertices
* improved robustness for holes in mesh
note: reducing the voxel size for laminar_coords_res usually fixes any residual holes

akhanf committed:
add rule for tanh normalization of surface metrics (#505)

- uses -5 to +5 as soft limits, only applied to native thickness,
curvature, gyrification
- fixes unfoldreg failures when extreme values exist

akhanf committed:
enforces connected-components in src/sink labels (#507)

Additional step in postproc_boundary_vertices for spatial consistency of each label, by keeping the largest connected component, and relabelling smaller components based on their neighbours.

akhanf committed:
updates dev dependencies (#508)

pinned pygraphviz was creating issues..

akhanf committed:
Hole fill radius update, hole reporting in get_boundary_vertices (#509)

* update hole fill radius
* paint holes a different color when calculating boundary vertices

also reports info on the holes in the log output

akhanf committed:
adds a small amount of tikhonov regularization (#510)

should fix any remaining solver issues (cause by holes or other
unnoticeable mesh defects)

Dhananjhay committed:
use action to setup and register a runner

Dhananjhay committed:
correct the name of the branch

Dhananjhay committed:
supply arbutus credentials as input parameters

Dhananjhay committed:
export runner_pat

Dhananjhay committed:
stop exporting credentials as inputs to action

Dhananjhay committed:
capture the vm_id output

Dhananjhay committed:
fix output directive

Dhananjhay committed:
change name of the runner from runner-pr-# to repo_name-pr-#

Dhananjhay committed:
use modularized remove runner github action

Dhananjhay committed:
use workflow instead of actions

Dhananjhay committed:
change .yaml to .yml

Dhananjhay committed:
add an option to manually trigger the wet run

Dhananjhay committed:
remove output directive

Dhananjhay committed:
remove id directive

Dhananjhay committed:
pass secrets

Dhananjhay committed:
stop inputting repository_name/owner

Dhananjhay committed:
do some clean up and add an action to setup snakebids env

Dhananjhay committed:
remove unwanted env variables and use an action for downloading wet run test data

Dhananjhay committed:
change labels

Dhananjhay committed:
prevent concurrent wet run tests for a PR

Dhananjhay committed:
add comment for downloading wet run test data

Dhananjhay committed:
trigger CI

Dhananjhay committed:
trigger CI

Dhananjhay committed:
trigger CI

Dhananjhay committed:
trigger CI

Dhananjhay committed:
set conda_pkgs_dir env var

Dhananjhay committed:
add issue and pr templates

Dhananjhay committed:
change from ANTs to Hippunfold

Dhananjhay committed:
set conda_pkgs_dirs as an env

Dhananjhay committed:
remove cloud-init.yml

Dhananjhay committed:
add a side by side comparison of qc plots

Dhananjhay committed:
fix path to lowresMRI data

Dhananjhay committed:
add permission to write on PRs

Dhananjhay committed:
lint

akhanf reviewed (changes requested):
This generally looks good to me! The commit history is a bit messy so maybe could use a rebase to clean-up?

Dhananjhay committed:
remove socket logic and use git

Dhananjhay committed:
update error messages

Dhananjhay committed:
call khanlab/actions from main branch

Dhananjhay committed:
run wet run test only when a label is attached

Dhananjhay committed:
add missing reg tag to khalab/actions

Dhananjhay commented:
> This generally looks good to me! The commit history is a bit messy so maybe could use a rebase to clean-up?

Hey @akhanf! Rebasing didn't help clean up the commit history, it's really messy 🥲 so instead I've created a new PR (#521) which was branched off from `dev-v2.0.0`. It's much easier to see the changes pertaining to wet run ephemeral test logic now. I can close this PR once you approve!
